File: sdk/kotlin/src/main/kotlin/com/privacyrpc/sdk/DAppDetector.kt
package com.privacyrpc.sdk

/**
 * dApp & Wallet Detector for PrivacyRPC
 *
 * Detects known Solana dApps and crypto wallets, matching the
 * Chrome extension's detection databases.
 */
class DAppDetector {

    // ── Data Models ──────────────────────────────────────────

    data class DAppInfo(
        val hostname: String,
        val name: String,
        val category: DAppCategory,
        val isTrusted: Boolean = true
    )

    enum class DAppCategory {
        DEX, LENDING, NFT, WALLET, EXPLORER, INFRASTRUCTURE, BRIDGE, GOVERNANCE, OTHER
    }

    data class WalletInfo(
        val id: String,
        val name: String,
        val type: String = "wallet"
    )

    data class SiteAnalysis(
        val hostname: String,
        val isDApp: Boolean,
        val dAppInfo: DAppInfo?,
        val isWalletSite: Boolean,
        val category: DAppCategory?
    )

    // ── Known Solana dApps (matching extension popup.js) ─────

    companion object {
        val KNOWN_DAPPS: Map<String, DAppInfo> = mapOf(
            // DEXs
            "jup.ag" to DAppInfo("jup.ag", "Jupiter", DAppCategory.DEX),
            "jupiter.ag" to DAppInfo("jupiter.ag", "Jupiter", DAppCategory.DEX),
            "raydium.io" to DAppInfo("raydium.io", "Raydium", DAppCategory.DEX),
            "orca.so" to DAppInfo("orca.so", "Orca", DAppCategory.DEX),
            "lifinity.io" to DAppInfo("lifinity.io", "Lifinity", DAppCategory.DEX),
            "meteora.ag" to DAppInfo("meteora.ag", "Meteora", DAppCategory.DEX),
            "phoenix.trade" to DAppInfo("phoenix.trade", "Phoenix", DAppCategory.DEX),
            "drift.trade" to DAppInfo("drift.trade", "Drift", DAppCategory.DEX),
            "zeta.markets" to DAppInfo("zeta.markets", "Zeta Markets", DAppCategory.DEX),

            // Lending / DeFi
            "marinade.finance" to DAppInfo("marinade.finance", "Marinade", DAppCategory.LENDING),
            "solend.fi" to DAppInfo("solend.fi", "Solend", DAppCategory.LENDING),
            "mango.markets" to DAppInfo("mango.markets", "Mango Markets", DAppCategory.LENDING),
            "kamino.finance" to DAppInfo("kamino.finance", "Kamino", DAppCategory.LENDING),
            "marginfi.com" to DAppInfo("marginfi.com", "marginfi", DAppCategory.LENDING),
            "solblaze.org" to DAppInfo("solblaze.org", "SolBlaze", DAppCategory.LENDING),
            "jito.network" to DAppInfo("jito.network", "Jito", DAppCategory.LENDING),

            // NFT Marketplaces
            "magiceden.io" to DAppInfo("magiceden.io", "Magic Eden", DAppCategory.NFT),
            "tensor.trade" to DAppInfo("tensor.trade", "Tensor", DAppCategory.NFT),
            "hyperspace.xyz" to DAppInfo("hyperspace.xyz", "Hyperspace", DAppCategory.NFT),
            "exchange.art" to DAppInfo("exchange.art", "Exchange Art", DAppCategory.NFT),
            "formfunction.xyz" to DAppInfo("formfunction.xyz", "Formfunction", DAppCategory.NFT),
            "solanart.io" to DAppInfo("solanart.io", "Solanart", DAppCategory.NFT),
            "opensea.io" to DAppInfo("opensea.io", "OpenSea", DAppCategory.NFT),

            // Wallets
            "phantom.app" to DAppInfo("phantom.app", "Phantom", DAppCategory.WALLET),
            "solflare.com" to DAppInfo("solflare.com", "Solflare", DAppCategory.WALLET),
            "backpack.app" to DAppInfo("backpack.app", "Backpack", DAppCategory.WALLET),
            "glow.app" to DAppInfo("glow.app", "Glow", DAppCategory.WALLET),

            // Explorers
            "solana.com" to DAppInfo("solana.com", "Solana", DAppCategory.EXPLORER),
            "solscan.io" to DAppInfo("solscan.io", "Solscan", DAppCategory.EXPLORER),
            "solanabeach.io" to DAppInfo("solanabeach.io", "Solana Beach", DAppCategory.EXPLORER),
            "explorer.solana.com" to DAppInfo("explorer.solana.com", "Solana Explorer", DAppCategory.EXPLORER),
            "xray.helius.xyz" to DAppInfo("xray.helius.xyz", "XRAY", DAppCategory.EXPLORER),
            "solana.fm" to DAppInfo("solana.fm", "SolanaFM", DAppCategory.EXPLORER),

            // Infrastructure
            "squads.so" to DAppInfo("squads.so", "Squads", DAppCategory.GOVERNANCE),
            "realms.today" to DAppInfo("realms.today", "Realms", DAppCategory.GOVERNANCE),
            "dialect.to" to DAppInfo("dialect.to", "Dialect", DAppCategory.INFRASTRUCTURE),
            "helius.dev" to DAppInfo("helius.dev", "Helius", DAppCategory.INFRASTRUCTURE),
            "shyft.to" to DAppInfo("shyft.to", "Shyft", DAppCategory.INFRASTRUCTURE),
            "quicknode.com" to DAppInfo("quicknode.com", "QuickNode", DAppCategory.INFRASTRUCTURE),
            "alchemy.com" to DAppInfo("alchemy.com", "Alchemy", DAppCategory.INFRASTRUCTURE)
        )

        /** Known wallet extension IDs (matching extension background.js) */
        val KNOWN_WALLETS: Map<String, WalletInfo> = mapOf(
            // Phantom
            "bfnaelmomeimhlpmgjnjophhpkkoljpa" to WalletInfo("bfnaelmomeimhlpmgjnjophhpkkoljpa", "Phantom"),
            "gojhcdgcpbpfigcaejpfhfegekdlneif" to WalletInfo("gojhcdgcpbpfigcaejpfhfegekdlneif", "Phantom (Dev)"),
            // Backpack
            "aflkmfhebedbjioipglgcbcmnbpgliof" to WalletInfo("aflkmfhebedbjioipglgcbcmnbpgliof", "Backpack"),
            "jnlgamecbpmbajjfhmmmlhejkemejdma" to WalletInfo("jnlgamecbpmbajjfhmmmlhejkemejdma", "Backpack"),
            // Solflare
            "bhhhlbepdkbapadjdnnojkbgioiodbic" to WalletInfo("bhhhlbepdkbapadjdnnojkbgioiodbic", "Solflare"),
            // MetaMask
            "nkbihfbeogaeaoehlefnkodbefgpgknn" to WalletInfo("nkbihfbeogaeaoehlefnkodbefgpgknn", "MetaMask"),
            "ejbalbakoplchlghecdalmeeeajnimhm" to WalletInfo("ejbalbakoplchlghecdalmeeeajnimhm", "MetaMask (Edge)"),
            // OKX
            "mcohilncbfahbmgdjkbpemcciiolgcge" to WalletInfo("mcohilncbfahbmgdjkbpemcciiolgcge", "OKX Wallet"),
            // Binance
            "fhbohimaelbohpjbbldcngcnapndodjp" to WalletInfo("fhbohimaelbohpjbbldcngcnapndodjp", "Binance Wallet"),
            // Glow
            "cfadjkfokiepapnlpbpdmaeajnhheghf" to WalletInfo("cfadjkfokiepapnlpbpdmaeajnhheghf", "Glow"),
            // Coinbase
            "dlcobpjiigpikoobohmabehhmhfoodbb" to WalletInfo("dlcobpjiigpikoobohmabehhmhfoodbb", "Coinbase Wallet"),
            "hnfanknocfeofbddgcijnmhnfnkdnaad" to WalletInfo("hnfanknocfeofbddgcijnmhnfnkdnaad", "Coinbase Wallet (Dev)"),
            // Slope
            "pocmplpaccanhmnllbbkpgfliimjljgo" to WalletInfo("pocmplpaccanhmnllbbkpgfliimjljgo", "Slope"),
            // Trust Wallet
            "ibnejdfjmmkpcnlpebklmnkoeoihofec" to WalletInfo("ibnejdfjmmkpcnlpebklmnkoeoihofec", "Trust Wallet"),
            "egjidjbpglichdcondbcbdnbeeppgdph" to WalletInfo("egjidjbpglichdcondbcbdnbeeppgdph", "Trust Wallet"),
            // Exodus
            "aholpfdialjgjfhomihkjbmgjidlcdno" to WalletInfo("aholpfdialjgjfhomihkjbmgjidlcdno", "Exodus"),
            // Rabby
            "acmacodkjbdgmoleebolmdjonilkdbch" to WalletInfo("acmacodkjbdgmoleebolmdjonilkdbch", "Rabby Wallet"),
            // Keplr
            "dmkamcknogkgcdfhhbddcghachkejeap" to WalletInfo("dmkamcknogkgcdfhhbddcghachkejeap", "Keplr"),
            // Leap
            "fcfcfllfndlomdhbehjjcoimbgofdncg" to WalletInfo("fcfcfllfndlomdhbehjjcoimbgofdncg", "Leap Wallet"),
            // Magic Eden
            "mkpegjkblkkefacfnmkajcjmabijhclg" to WalletInfo("mkpegjkblkkefacfnmkajcjmabijhclg", "Magic Eden Wallet"),
            // TipLink
            "gfkepgoophebjcgfkfgjbdkfgfcndbag" to WalletInfo("gfkepgoophebjcgfkfgjbdkfgfcndbag", "TipLink Wallet")
        )

        /** Trusted RPC endpoints (matching extension popup.js) */
        val TRUSTED_RPC_ENDPOINTS = listOf(
            "api.mainnet-beta.solana.com",
            "api.devnet.solana.com",
            "api.testnet.solana.com",
            "solana-api.projectserum.com",
            "rpc.helius.xyz",
            "mainnet.helius-rpc.com",
            "solana-mainnet.g.alchemy.com",
            "solana-mainnet.quiknode.pro",
            "ssc-dao.genesysgo.net"
        )

        /** Wallet-related name keywords */
        private val WALLET_KEYWORDS = listOf(
            "wallet", "phantom", "solana", "crypto", "backpack",
            "solflare", "metamask", "coinbase", "trust", "ledger",
            "trezor", "defi", "ethereum", "web3"
        )
    }

    // ── Detection Methods ────────────────────────────────────

    /** Check if a hostname belongs to a known dApp */
    fun isDApp(hostname: String): Boolean {
        return KNOWN_DAPPS.any { hostname.contains(it.key) }
    }

    /** Get dApp info for a hostname */
    fun getDAppInfo(hostname: String): DAppInfo? {
        return KNOWN_DAPPS.entries.firstOrNull { hostname.contains(it.key) }?.value
    }

    /** Analyze a site and return structured info */
    fun analyzeSite(hostname: String): SiteAnalysis {
        val dAppInfo = getDAppInfo(hostname)
        return SiteAnalysis(
            hostname = hostname,
            isDApp = dAppInfo != null,
            dAppInfo = dAppInfo,
            isWalletSite = dAppInfo?.category == DAppCategory.WALLET,
            category = dAppInfo?.category
        )
    }

    /** Check if a wallet extension ID is known */
    fun isKnownWallet(extensionId: String): Boolean {
        return extensionId in KNOWN_WALLETS
    }

    /** Get wallet info by extension ID */
    fun getWalletInfo(extensionId: String): WalletInfo? {
        return KNOWN_WALLETS[extensionId]
    }

    /** Check if a name looks like a wallet extension */
    fun isWalletByName(name: String): Boolean {
        val lower = name.lowercase()
        return WALLET_KEYWORDS.any { lower.contains(it) }
    }

    /** Check if an RPC endpoint is trusted */
    fun isTrustedEndpoint(url: String): Boolean {
        return TRUSTED_RPC_ENDPOINTS.any { url.contains(it) }
    }
}

File: sdk/swift/Sources/PrivacyRPCSDK/DAppDetector.swift
import Foundation

/// dApp & Wallet Detector for PrivacyRPC
///
/// Detects known Solana dApps and crypto wallets.
/// Matches the Chrome extension's detection databases.
public class DAppDetector {

    // MARK: - Models

    public struct DAppInfo {
        public let hostname: String
        public let name: String
        public let category: DAppCategory
        public let isTrusted: Bool

        public init(hostname: String, name: String, category: DAppCategory, isTrusted: Bool = true) {
            self.hostname = hostname
            self.name = name
            self.category = category
            self.isTrusted = isTrusted
        }
    }

    public enum DAppCategory: String {
        case dex = "DEX"
        case lending = "LENDING"
        case nft = "NFT"
        case wallet = "WALLET"
        case explorer = "EXPLORER"
        case infrastructure = "INFRASTRUCTURE"
        case bridge = "BRIDGE"
        case governance = "GOVERNANCE"
        case other = "OTHER"
    }

    public struct WalletInfo {
        public let id: String
        public let name: String
    }

    public struct SiteAnalysis {
        public let hostname: String
        public let isDApp: Bool
        public let dAppInfo: DAppInfo?
        public let isWalletSite: Bool
        public let category: DAppCategory?
    }

    // MARK: - Known Solana dApps (matching extension popup.js)

    public static let knownDApps: [String: DAppInfo] = {
        var dapps: [String: DAppInfo] = [:]
        let entries: [(String, String, DAppCategory)] = [
            // DEXs
            ("jup.ag", "Jupiter", .dex),
            ("jupiter.ag", "Jupiter", .dex),
            ("raydium.io", "Raydium", .dex),
            ("orca.so", "Orca", .dex),
            ("lifinity.io", "Lifinity", .dex),
            ("meteora.ag", "Meteora", .dex),
            ("phoenix.trade", "Phoenix", .dex),
            ("drift.trade", "Drift", .dex),
            ("zeta.markets", "Zeta Markets", .dex),

            // Lending / DeFi
            ("marinade.finance", "Marinade", .lending),
            ("solend.fi", "Solend", .lending),
            ("mango.markets", "Mango Markets", .lending),
            ("kamino.finance", "Kamino", .lending),
            ("marginfi.com", "marginfi", .lending),
            ("solblaze.org", "SolBlaze", .lending),
            ("jito.network", "Jito", .lending),

            // NFT Marketplaces
            ("magiceden.io", "Magic Eden", .nft),
            ("tensor.trade", "Tensor", .nft),
            ("hyperspace.xyz", "Hyperspace", .nft),
            ("exchange.art", "Exchange Art", .nft),
            ("formfunction.xyz", "Formfunction", .nft),
            ("solanart.io", "Solanart", .nft),
            ("opensea.io", "OpenSea", .nft),

            // Wallets
            ("phantom.app", "Phantom", .wallet),
            ("solflare.com", "Solflare", .wallet),
            ("backpack.app", "Backpack", .wallet),
            ("glow.app", "Glow", .wallet),

            // Explorers
            ("solana.com", "Solana", .explorer),
            ("solscan.io", "Solscan", .explorer),
            ("solanabeach.io", "Solana Beach", .explorer),
            ("explorer.solana.com", "Solana Explorer", .explorer),
            ("xray.helius.xyz", "XRAY", .explorer),
            ("solana.fm", "SolanaFM", .explorer),

            // Infrastructure
            ("squads.so", "Squads", .governance),
            ("realms.today", "Realms", .governance),
            ("dialect.to", "Dialect", .infrastructure),
            ("helius.dev", "Helius", .infrastructure),
            ("shyft.to", "Shyft", .infrastructure),
            ("quicknode.com", "QuickNode", .infrastructure),
            ("alchemy.com", "Alchemy", .infrastructure)
        ]

        for (host, name, cat) in entries {
            dapps[host] = DAppInfo(hostname: host, name: name, category: cat)
        }
        return dapps
    }()

    // MARK: - Known Wallet Extension IDs (matching extension background.js)

    public static let knownWallets: [String: WalletInfo] = [
        "bfnaelmomeimhlpmgjnjophhpkkoljpa": WalletInfo(id: "bfnaelmomeimhlpmgjnjophhpkkoljpa", name: "Phantom"),
        "gojhcdgcpbpfigcaejpfhfegekdlneif": WalletInfo(id: "gojhcdgcpbpfigcaejpfhfegekdlneif", name: "Phantom (Dev)"),
        "aflkmfhebedbjioipglgcbcmnbpgliof": WalletInfo(id: "aflkmfhebedbjioipglgcbcmnbpgliof", name: "Backpack"),
        "jnlgamecbpmbajjfhmmmlhejkemejdma": WalletInfo(id: "jnlgamecbpmbajjfhmmmlhejkemejdma", name: "Backpack"),
        "bhhhlbepdkbapadjdnnojkbgioiodbic": WalletInfo(id: "bhhhlbepdkbapadjdnnojkbgioiodbic", name: "Solflare"),
        "nkbihfbeogaeaoehlefnkodbefgpgknn": WalletInfo(id: "nkbihfbeogaeaoehlefnkodbefgpgknn", name: "MetaMask"),
        "ejbalbakoplchlghecdalmeeeajnimhm": WalletInfo(id: "ejbalbakoplchlghecdalmeeeajnimhm", name: "MetaMask (Edge)"),
        "mcohilncbfahbmgdjkbpemcciiolgcge": WalletInfo(id: "mcohilncbfahbmgdjkbpemcciiolgcge", name: "OKX Wallet"),
        "fhbohimaelbohpjbbldcngcnapndodjp": WalletInfo(id: "fhbohimaelbohpjbbldcngcnapndodjp", name: "Binance Wallet"),
        "cfadjkfokiepapnlpbpdmaeajnhheghf": WalletInfo(id: "cfadjkfokiepapnlpbpdmaeajnhheghf", name: "Glow"),
        "dlcobpjiigpikoobohmabehhmhfoodbb": WalletInfo(id: "dlcobpjiigpikoobohmabehhmhfoodbb", name: "Coinbase Wallet"),
        "hnfanknocfeofbddgcijnmhnfnkdnaad": WalletInfo(id: "hnfanknocfeofbddgcijnmhnfnkdnaad", name: "Coinbase Wallet (Dev)"),
        "pocmplpaccanhmnllbbkpgfliimjljgo": WalletInfo(id: "pocmplpaccanhmnllbbkpgfliimjljgo", name: "Slope"),
        "ibnejdfjmmkpcnlpebklmnkoeoihofec": WalletInfo(id: "ibnejdfjmmkpcnlpebklmnkoeoihofec", name: "Trust Wallet"),
        "egjidjbpglichdcondbcbdnbeeppgdph": WalletInfo(id: "egjidjbpglichdcondbcbdnbeeppgdph", name: "Trust Wallet"),
        "aholpfdialjgjfhomihkjbmgjidlcdno": WalletInfo(id: "aholpfdialjgjfhomihkjbmgjidlcdno", name: "Exodus"),
        "acmacodkjbdgmoleebolmdjonilkdbch": WalletInfo(id: "acmacodkjbdgmoleebolmdjonilkdbch", name: "Rabby Wallet"),
        "dmkamcknogkgcdfhhbddcghachkejeap": WalletInfo(id: "dmkamcknogkgcdfhhbddcghachkejeap", name: "Keplr"),
        "fcfcfllfndlomdhbehjjcoimbgofdncg": WalletInfo(id: "fcfcfllfndlomdhbehjjcoimbgofdncg", name: "Leap Wallet"),
        "mkpegjkblkkefacfnmkajcjmabijhclg": WalletInfo(id: "mkpegjkblkkefacfnmkajcjmabijhclg", name: "Magic Eden Wallet"),
        "gfkepgoophebjcgfkfgjbdkfgfcndbag": WalletInfo(id: "gfkepgoophebjcgfkfgjbdkfgfcndbag", name: "TipLink Wallet")
    ]

    // MARK: - Trusted RPC Endpoints

    public static let trustedRpcEndpoints: [String] = [
        "api.mainnet-beta.solana.com",
        "api.devnet.solana.com",
        "api.testnet.solana.com",
        "solana-api.projectserum.com",
        "rpc.helius.xyz",
        "mainnet.helius-rpc.com",
        "solana-mainnet.g.alchemy.com",
        "solana-mainnet.quiknode.pro",
        "ssc-dao.genesysgo.net"
    ]

    private static let walletKeywords = [
        "wallet", "phantom", "solana", "crypto", "backpack",
        "solflare", "metamask", "coinbase", "trust", "ledger",
        "trezor", "defi", "ethereum", "web3"
    ]

    public init() {}

    // MARK: - Detection Methods

    /// Check if a hostname belongs to a known dApp
    public func isDApp(_ hostname: String) -> Bool {
        return Self.knownDApps.contains { hostname.contains($0.key) }
    }

    /// Get dApp info for a hostname
    public func getDAppInfo(_ hostname: String) -> DAppInfo? {
        return Self.knownDApps.first { hostname.contains($0.key) }?.value
    }

    /// Analyze a site and return structured info
    public func analyzeSite(_ hostname: String) -> SiteAnalysis {
        let dAppInfo = getDAppInfo(hostname)
        return SiteAnalysis(
            hostname: hostname,
            isDApp: dAppInfo != nil,
            dAppInfo: dAppInfo,
            isWalletSite: dAppInfo?.category == .wallet,
            category: dAppInfo?.category
        )
    }

    /// Check if a wallet extension ID is known
    public func isKnownWallet(_ extensionId: String) -> Bool {
        return Self.knownWallets[extensionId] != nil
    }

    /// Get wallet info by extension ID
    public func getWalletInfo(_ extensionId: String) -> WalletInfo? {
        return Self.knownWallets[extensionId]
    }

    /// Check if a name looks like a wallet extension
    public func isWalletByName(_ name: String) -> Bool {
        let lower = name.lowercased()
        return Self.walletKeywords.contains { lower.contains($0) }
    }

    /// Check if an RPC endpoint is trusted
    public func isTrustedEndpoint(_ url: String) -> Bool {
        return Self.trustedRpcEndpoints.contains { url.contains($0) }
    }
}
